### ISSUE TITLE
Option to not create router dynamically

### ISSUE THREAD

**Issue by Sarke** [containous/traefik#5576]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

In Traefik v2.0, I would like to be able to define a docker service using labels, but not have it create a router dynamically.  My use case is that the service will be used by several dynamically defined routers from a file.

A workaround is to define a useless route

```yaml
  labels:
    - "traefik.enable=true"
    - "traefik.http.routers.do-nothing.entrypoints=web,websecure"
    - "traefik.http.services.real-service.loadbalancer.server.port=80"
```

This creates a new router with a rule of something like ``Host(`stackname-servicename`)`` but this is just noise really, and not ideal.

**Comment by dduportal** [comment]
Hi @Sarke , thanks for your interest. Could you clarify the use case please? We are not sure to understand what you want to achieve when you mean `My use case is that the service will be used by several dynamically defined routers from a file.`?

**Comment by Sarke** [comment]
So I define several routers in a file.  This includes different host names and SSL certs, etc.  It doesn't really matter where this comes from, just that it is generated and is not in the labels for the docker service.

In order for these routers from this file to point to the docker service, this needs to be enabled in the labels, right?  So I can do `service = "real-service@docker"` in the toml file.

The other options would be to:

A) Define the service in the toml file as well.  But how would I do this?  Would I have to look up the service's IP and use that as the loadbalancer url?  It's much simpler to just enable it in the docker compose file, so I can refer to it with `real-service@docker`.

B) Have everything in the docker-compose labels.  I think this is what you're thinking, but the toml will be generated by an external script, and it is much easier than setting the labels dynamically and messing with adding and removing them.  Or can you even change labels of a swarm service without re-deploying it? It's much simpler to just define these routers in a toml.

So, that leaves me with defining the service in the labels, and the routers in the toml.  But I end up with an extra useless router created by Traefik.

**Comment by dduportal** [comment]
Thanks @Sarke for the clarification 👍

**Comment by sodabrew** [comment]
Does my proposal in https://github.com/containous/traefik/issues/6040 look like what you’d want to see for consul-services to have file-defined router rules?

**Comment by OverloadUT** [comment]
I have another use-case that would benefit from this feature:

I set up a simple nginx container for serving up the error pages for the Errors middleware. The Errors middleware will forward the request directly to a service, and does not need a router.

**Comment by artentica** [comment]
Some advancement on that feature ?

**Comment by reesericci** [comment]
I just want to be able to do this so I can define middlewares in labels on the reverse-proxy container without creating a router.

**Comment by cleboo** [comment]
Another use-case is an A/B testing setup, like the one described in this blog post: https://iximiuz.com/en/posts/traefik-canary-deployments-with-weighted-load-balancing/

Hiding the automatically created routers by assigning them the traefik endpoint or an impossible route feels a bit wrong 😅

**Comment by sirkjohannsen** [comment]
I just figured out that you can disable the defaultRoute by setting it to  `""` : 
```
    defaultRule = ""

```

It does lead to a nasty error message though :( 
```
{"level":"error","msg":"Undefined rule","routerName":"my-service-name","time":"2021-07-14T13:47:07+02:00"}
```

**Comment by DaneHoffman** [comment]
Yeah I've also become accustomed to defining services in `docker-compose` flags, while defining routers in a dynamic configuration file. Seems like a smoother way of doing things (because you can change routing rules without restarting the docker service), and yet, if you do this, you end up with a huge number of useless/potentially-insecure routers created by the docker provider cluttering the dashboard interface. 

Definitely seems like an unintended behavior. Would love to see a flag that disables automatic router generation.

**Comment by averon420** [comment]
hello, I also would love to define the service in the docker-compose file and have everything else in dynamic configuration.
Can someone explain me how to use the "defaultRule" workarround? I don't really understand, tried alot.

**Comment by Wait4Code** [comment]
I'm interested too. Any workaround ?

**Comment by kitro** [comment]
> ```yaml
> services
> ```

That approach seems more practical, It will be good if traefik have a config to disable auto-route and specify the custom name of the container service.

**Comment by kitro** [comment]
> I've also recently started defining services in `docker-compose` flags, while defining routers in a dynamic configuration file. Seems like a more functional way of doing things, since you can change routing rules without restarting the docker service while still getting the benefits of automatic port assignment via the docker provider, and yet, if you do this, you end up with a huge number of useless/potentially-insecure routers created by the docker provider cluttering the dashboard interface.
> 
> Seems almost like an unintended behavior. Would love to see a flag that disables automatic router generation.

That approach seems more practical, It will be good if traefik have a config to disable auto-route and specify the custom name of the container service.

**Comment by tigerblue77** [comment]
> I have another use-case that would benefit from this feature:
> 
> I set up a simple nginx container for serving up the error pages for the Errors middleware. The Errors middleware will forward the request directly to a service, and does not need a router.

I'm in this exact use case, everything is working fine but I get this error message in logs :
> 2024-01-06T21:03:19+01:00 ERR | ACME CA=https://acme-v02.api.letsencrypt.org/directory msg=Unable to obtain ACME certificate for domains "traefik-http-errors-handler": unable to generate a certificate for the domains [traefik-http-errors-handler]: acme: error: 400 :: POST :: https://acme-v02.api.letsencrypt.org/acme/new-order :: urn:ietf:params:acme:error:rejectedIdentifier :: Error creating new order :: Cannot issue for "traefik-http-errors-handler": Domain name needs at least one dot providerName=letsencrypt.acme routerName=HTTP_error_pages@docker rule=Host(`Traefik-HTTP-errors-handler`)

Can someone confirm it's linked ?

**Comment by giftkugel** [comment]
I am using Traefik + Docker with  `docker.io/thomseddon/traefik-forward-auth` as middleware with the following configuration:

```yaml
   labels:
      traefik.enable: true
      traefik.http.middlewares.traefik-forward-auth.forwardauth.address: "http://traefik-forward-auth:4181"
      traefik.http.middlewares.traefik-forward-auth.forwardauth.authResponseHeaders: "X-Forwarded-User"
      traefik.http.middlewares.traefik-forward-auth.forwardauth.trustForwardHeader: true
      traefik.http.services.traefik-forward-auth.loadbalancer.server.port: 4181
```

Unfortunately Traefik creates a router with `Host('traefik-forward-auth')` and binds it to all my entrypoints ... even the ones it should not use at all.

Does a middleware really need a router, as I do not expect it to be accessed from outside.

I would love to disable that! Because otherwise it means that the middleware can be triggered from the outside with the a `Host: traefik-forward-auth` based request, right?

**Comment by marekhanzlik** [comment]
Hi, i have another usecase.
I use Traefik Nomad discovery - and services exposed in nomad (with `traefik.enable=true` flag) automatically create a routers for the services. 
We dont need them because we have routers created statically in file configuration (for reasons) and now traefik always wants to generate ACME certs for services that does not have a valid domain name (like `Host(service-A)`) and we are getting 429 throttled from the ACME side because we create too many invalid requests

It would be great to have a flag/label that either "resets" the automaticaly created router to nothing or disables this behaviour alotgether

**Comment by hotkeeper** [comment]
Are you going to implement this feature?
We have the use case that we have docker provider but are not using the docker labels to configure we are using external configuration YML files to define the routing and middleware, it makes the configuration much cleaner and simpler.
But we should get rid off the default routers `Host('stackname-servicename')` created for every service!

**Comment by yvess** [comment]
I have a similar use case than @hotkeeper. I just define the Traefik service through docker swarm, and want to use the dynamic configuration for multiple domains to pointing to the service. 
Currently I just define a "dummy" route which I don't use, just to prevent the creation of the default route. 
The nice thing about dynamic configuration is that I don't need to re-deploy the docker stack for each small change, and the toml config is more readable than docker labels.

One use case I use is that when I want to change a service, I can point all matching routers to a temporary maintenance service just with a simple very fast sed command.

**Comment by emilevauge** [comment]
Hello!

Thanks for the feedback, we've updated the status to `enhancement`.

While we think this is valuable, it won't make our roadmap soon as we're focused elsewhere. If you or another community member would like to build this, we'd love to support you! Let us know, and we'll help ensure you have all the information needed.

Make sure to check out our [contributor docs](https://doc.traefik.io/traefik/contributing/submitting-pull-requests/) and link your PR to this issue.

**Comment by nehart** [comment]
Hi,
What is the status of this feature? I would be interested too.
I would like to use Docker labels to define the service, which would then be referenced by the file provider.

BR,
Norli.
